### PR TITLE
CN Testing Page Quiz Functionality

### PR DESCRIPTION
**Pull request by chrisn02**
The user is now able to start and complete a quiz that has been assigned to them. It will be displayed in the "Tests to Complete" section and once clicked the quiz will begin. 
image
image The questions will be loaded. The correct answer is stored server side rather than client side to prevent users from using inspect element to find the correct answer
image When the user selects an answer a button will appear to go to the next question.

image The user has now advanced to the next question. There is also a progress bar that goes up with each question

image If there is a question with less than 4 answers, then only the needed number of boxes will appear

image Once the user has completed all questions a popup will appear saying that quiz is complete and there will be a button to view the results

image Another popup will appear with the results and these results will be saved to the database

image The quiz that was just completed has been moved to the "Completed Tests" section

### CODE REVIEW

File: pages/student/testing-page.php [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    //Get Quiz Set
+    $stmt = $conn->prepare("CALL GetAllQuestions(:quizID)");
+    $stmt->bindParam(":quizID", $quizID, PDO::PARAM_STR);
+    $stmt->execute();
+    $allQuestions = $stmt->fetchAll();
+
+    //Get question number and total number of questions
+    $questionNumber = 0;
+    $totalQuestions = count($allQuestions);
+}
```
Comment (Jforjo) [inline pages/student/testing-page.php:85]: I would heavily suggest using `GetQuizSet` to get a set a questions, as this randomises the order and allows you to limit how many questions are in each quiz. However, what you have now is fine.

File: pages/student/testing-page.php [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        <h1><?php echo htmlspecialchars($subjectName, ENT_QUOTES, 'UTF-8'); ?></h1>
+        <h2><?php echo htmlspecialchars($quizName, ENT_QUOTES, 'UTF-8'); ?></h2>
+    </div>
+    </div>
+    <div class="main-content">
+        <!-- Loop through each question -->
+        <?php foreach ($allQuestions as $question): ?>
+        <!-- Increment the question number -->
+        <?php $questionNumber++; ?>
+        <?php $correctAnswer = intval($question['correctAnswer']);?>
```
Comment (Jforjo) [inline pages/student/testing-page.php]: I don't see the purpose of storing the correct answer here.

File: pages/student/testing-page.php [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    </div>
+</div>
+
+    <div id="results-modal" class ="results-modal" style="display: none;">
+    <div class="results-content">
+        <h1>Results</h1>
+        <p id="total-score"><i class="fas fa-trophy"></i>Points Earned: </p>
+        <p id="correct-answers"><i class="fas fa-check"></i>Correct questions: </p>
+        <p id="percentage"><i class="fas fa-chart-line"></i>Percentage: </p>
+        <button onclick="window.location.href='index.php'">Finish</button>
```
Comment (Jforjo) [inline pages/student/testing-page.php]: Just use an `<a>` tag.

File: pages/student/testing-page.php [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+function checkAnswerAndUpdateProgress() {
+    var xhr = new XMLHttpRequest();
+    xhr.open("POST", "../../php/checkAnswer.php", true);
+    xhr.setRequestHeader("Content-Type", "application/x-www-form-urlencoded");
+    xhr.send(`quizID=${quizID}&questionNumber=${currentQuestion}&userChoice=${userChoice}&questionID=${questionID}`);
+    xhr.onreadystatechange = function() {
+        if (this.readyState == 4 && this.status == 200) {
+            // The server has responded
+            var response = JSON.parse(this.responseText);
+            console.log(this.responseText)
```
Comment (Jforjo) [inline pages/student/testing-page.php]: I think this was for testing purposes, so it can probably be removed now.

File: php/checkAnswer.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+$quizID = $_POST['quizID'];
+$questionID = $_POST['questionID'];
+$questionNumber = $_POST['questionNumber'];
+$userChoice = $_POST['userChoice'];
+
+$conn = newConn();
+
+// Get the correct answer from the database
+$stmt = $conn->prepare("SELECT correctAnswer FROM questions WHERE questionID = :questionID");
```
Comment (Jforjo) [inline php/checkAnswer.php]: Make a prepared statement for this.

File: php/updateCompletedQuiz.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    $correctCount = $_POST['correctCount'];
+    $points = $_POST['points'];
+    $dateCompleted = date("Y-m-d H:i:s");
+
+    $conn = newConn();
+    $sql = "
+        UPDATE `studentQuizLink`
+        SET `completed` = :completed, `questionCount` = :questionCount, `correctCount` = :correctCount, `points` = :points, `dateCompleted` = :dateCompleted
+        WHERE `studentQuizLinkID` = :studentQuizLinkID
+    ";
```
Comment (Jforjo) [inline php/updateCompletedQuiz.php]: Make a prepared statement for this.

**Review by Jforjo — CHANGES_REQUESTED** [review]
I think you should also start adding checks to your files - only let students access the index, testing-page, checkanswer, updatecompletedquiz and any other that only a student needs. Also, redirect them back to the login screen if they are not logged in.
See comments for more.

File: pages/student/testing-page.php (hunk 1 above)
Comment (chrisn02) [reply pages/student/testing-page.php:85]: I originally did use GetQuizSet, but I was having an error with it the other day so I switched to GetAllQuestions. Have switched back and it now works

File: pages/student/testing-page.php (hunk 2 above)
Comment (chrisn02) [reply pages/student/testing-page.php]: That was a leftover from when I was storing the correct answer client-side

File: php/checkAnswer.php (hunk 5 above)
Comment (chrisn02) [reply php/checkAnswer.php]: Could you clarify this please, I think it's already in a prepared statement but correct me if wrong

Comment (Jforjo) [reply php/checkAnswer.php]: I meant a prepared procedure/function.
Your SQL should look like "CALL myFunction();" and not what you have now.

Comment (chrisn02) [reply php/checkAnswer.php]: Ah, got it. Will do that now

**Comment by chrisn02** [comment]
Have now added checks to only allow the students to access the necessary pages and a check to redirect the user to the login page if they are not logged in. Also used a procedure for the SQL queries

File: pages/student/index.php [hunk 7]
```diff
@@ -20,11 +20,25 @@
     </header>
     <?php 
     require_once('../../php/connection.php');
+    require_once(__DIR__ . '/../../php/dbfuncs.php');
```
Comment (Jforjo) [inline pages/student/index.php:22]: The connection file is required in the dbfuncs file, so you only need the dbfuncs require line.

Comment (chrisn02) [reply pages/student/index.php:22]: Sorted

**Review by Jforjo — APPROVED** [review]

**Review by gchodancar19 — APPROVED** [review]
I think the fields created looks successful as it shows when the next field is selected and also shows the progress of when of how much question have been completed and at last shows the test results.

**Pull request merged**